provider/virtio_mem_utils.py

```diff
@@ -81,19 +96,53 @@ def check_memory_devices(device_id, cfg_req_size_param, threshold, vm, test):
             size = int(data.get("size"))
             node = int(data.get("node"))
             requested_size = int(data.get("requested-size"))
-            error_context.context("node: %d, id: %s, size: %d and requested-size: %d"
-                                  % (node, id, size, requested_size), test.log.debug)
+            error_context.context(
+                "node: %d, id: %s, size: %d and requested-size: %d"
+                % (node, id, size, requested_size),
+                test.log.debug,
+            )
             node_size = float(get_node_size(node, vm, test))
-            normalized_size = float(normalize_data_size(str(size), 'M'))
+            normalized_size = float(normalize_data_size(str(size), "M"))
             mem_limit = (node_size - normalized_size) * threshold
             if size != requested_size and normalized_size > mem_limit:
-                test.fail("Error, size: %d and requested-size: %d are not equals."
-                          % (size, requested_size))
+                test.fail(
+                    "Error, size: %d and requested-size: %d are not equals."
+                    % (size, requested_size)
+                )
             if cfg_req_size_param != "0":
-                req_size_param = int(float(normalize_data_size(cfg_req_size_param, 'B')))
+                req_size_param = int(
+                    float(normalize_data_size(cfg_req_size_param, "B"))
+                )
             else:
                 req_size_param = int(cfg_req_size_param)
             error_context.context("req_size_param: %d" % req_size_param, test.log.debug)
             if requested_size != req_size_param:
-                test.fail("Error, requested-size: %d is not the specified in the cfg: %d."
-                          % (requested_size, req_size_param))
+                test.fail(
+                    "Error, requested-size: %d is not the specified in the cfg: %d."
+                    % (requested_size, req_size_param)
+                )
+
+
+def count_memslots(vm):
+    """
+    Returns the number of memslots assigned to the current virtio-mem device
+    :param vm: the VM object
+    """
+    output = vm.monitor.info("mtree").splitlines()
+    memslots = set()
+    for memory_region in output:
+        memory_region = memory_region.strip()
+        if "mem-vmem0" in memory_region and "memslot" in memory_region:
```

Comment:
We can not ensure the VM boot with `mem_devs = 'vmem0'` and `memdev_memory_vmem0 = "mem-vmem0"` since the values depend on the parameter in cfg by the user.
So it does not fit the general scenarios.

Reply:
@YongxueHong so receiving the memory object as an argument in the function could be enough?

Reply:
@YongxueHong code updated !